be/src/formats/parquet/level_builder.cpp

```diff
@@ -0,0 +1,656 @@
+                                               const ::parquet::schema::NodePtr& node, const ColumnPtr& col,
+                                               const CallbackFunction& write_leaf_callback) {
+    DCHECK(type_desc.type == TYPE_ARRAY);
+    auto outer_node = std::static_pointer_cast<::parquet::schema::GroupNode>(node);
+    auto mid_node = std::static_pointer_cast<::parquet::schema::GroupNode>(outer_node->field(0));
+    auto inner_node = mid_node->field(0);
+
+    const auto null_col = get_raw_null_column(col);
+    const auto array_col = down_cast<ArrayColumn*>(ColumnHelper::get_data_column(col.get()));
+    const auto& elements = array_col->elements_column();
+    const auto& offsets = array_col->offsets_column()->get_data();
+
+    auto def_levels = std::make_shared<std::vector<int16_t>>();
+    auto rep_levels = std::make_shared<std::vector<int16_t>>();
+
+    int offset = 0;
+    for (auto i = 0; i < ctx.num_levels(); i++) {
+        auto def_level = ctx._def_levels ? ctx._def_levels->at(i) : 0;
+        auto rep_level = ctx._rep_levels ? ctx._rep_levels->at(i) : 0;
+        if (def_level < ctx._max_def_level) {
+            def_levels->push_back(def_level);
+            rep_levels->push_back(rep_level);
+            continue;
+        }
+
+        if (null_col != nullptr && null_col[offset]) {
+            def_levels->push_back(def_level);
+            rep_levels->push_back(rep_level);
+
+            offset++;
+            continue;
+        }
+
+        auto array_size = offsets[offset + 1] - offsets[offset];
+        if (array_size == 0) {
+            def_levels->push_back(def_level + node->is_optional());
+            rep_levels->push_back(rep_level);
+        } else {
+            def_levels->push_back(def_level + node->is_optional() + 1);
+            rep_levels->push_back(rep_level);
+            for (auto j = 1; j < array_size; j++) {
+                def_levels->push_back(def_level + node->is_optional() + 1);
+                rep_levels->push_back(ctx._max_rep_level + 1);
+            }
+        }
+        offset++;
+    }
+
+    LevelBuilderContext derived_ctx(def_levels->size(), def_levels, ctx._max_def_level + node->is_optional() + 1,
+                                    rep_levels, ctx._max_rep_level + 1);
+
+    DCHECK(elements->size() == offset);
+    return _write_column_chunk(derived_ctx, type_desc.children[0], inner_node, elements, write_leaf_callback);
+}
+
+Status LevelBuilder::_write_array_column_chunkV2(const LevelBuilderContext& ctx, const TypeDescriptor& type_desc,
+                                                 const ::parquet::schema::NodePtr& node, const ColumnPtr& col,
+                                                 const CallbackFunction& write_leaf_callback) {
+    DCHECK(type_desc.type == TYPE_ARRAY);
+    auto outer_node = std::static_pointer_cast<::parquet::schema::GroupNode>(node);
```

Comment:
comment why there is three level for maintanance. otherwise, We need to check for the sepc of parquet to know why everytime.